src/unit-stack-configuration-workspace/unit-stack-configuration.service.ts

```diff
@@ -62,71 +62,169 @@ export class UnitStackConfigurationWorkspaceService {
     return await this.repository.getUnitStackConfigsByLocationIds(locationIds);
   }
 
-  runUnitStackChecks(monitorPlan: UpdateMonitorPlanDTO): string[] {
+  async runUnitStackChecks(monitorPlan: UpdateMonitorPlanDTO, facilityId: number): Promise<string[]> {
     const errorList: string[] = [];
 
-    // Check for duplicate unit stack configurations.
-    monitorPlan.unitStackConfigurationData.forEach((usc1, i) => {
-      if (!usc1.unitId) {
-        const error = CheckCatalogService.formatResultMessage('MONLOC-107-A', {
-          fieldname: 'unitId',
-          key: KEY
-        })
-        errorList.push(error)
-      } else if (
-        monitorPlan.unitStackConfigurationData.findIndex(
-          usc2 =>
-            usc1.unitId === usc2.unitId &&
-            usc1.stackPipeId === usc2.stackPipeId &&
-            usc1.beginDate === usc2.beginDate,
-        ) !== i
-      ) {
-        errorList.push(
-          `[MONLOC-107-B] Duplicate Unit Stack Configuration record found at index #${i}.`,
-        );
-      }
-    });
+    await this.validateUnitStackConfigs(monitorPlan, facilityId, errorList);
+    this.validateConfigLocationRelationships(monitorPlan, errorList);
+    this.validateStackUnitAssociations(monitorPlan, errorList);
 
-    const unitStackIds: Set<string> = new Set<string>(); // Set for faster look up times
-    const unitUnitIds: Set<string> = new Set<string>();
+    return errorList
+  }
 
-    const unitStackConfigStackIds: Set<string> = new Set<string>();
+  private async validateUnitStackConfigs(
+    monitorPlan: UpdateMonitorPlanDTO,
+    facilityId: number,
+    errorList: string[],
+  ): Promise<void> {
+    for (const unitStackConfig of monitorPlan.unitStackConfigurationData) {
+      this.validateUnitStackConfig(unitStackConfig, errorList);
 
-    for (const location of monitorPlan.monitoringLocationData) {
-      if (location.stackPipeId) {
-        unitStackIds.add(location.stackPipeId);
-      }
-      if (location.unitId) {
-        unitUnitIds.add(location.unitId);
+      if (unitStackConfig.unitId && unitStackConfig.stackPipeId) {
+        const hasDuplicate = await this.checkDatabaseDuplicate(
+          unitStackConfig,
+          facilityId,
+          errorList,
+        );
+        if (hasDuplicate) return;
       }
```

Comment:
It shouldn't be an error if there is already a USC record in the database. The user should be able to export then directly import a monitor plan, or they may be setting the end date on a USC record. I think the previous logic of making sure there aren't duplicate USC records _in the file_ should remain.